en/middleware/mavlink.md

```diff
@@ -2,200 +2,419 @@
+For example, you might add the following line to [/ROMFS/px4fmu_common/init.d-posix/px4-rc.mavlink](https://github.com/PX4/PX4-Autopilot/blob/main/ROMFS/px4fmu_common/init.d-posix/px4-rc.mavlink) in order to stream `BATTERY_STATUS_DEMO` at 50Hz on UDP port `14556` (`-r` configures the streaming rate and `-u` identifies the MAVLink channel on UDP port 14556).
+
+```sh
+mavlink stream -r 50 -s BATTERY_STATUS_DEMO -u 14556
 ```
 
-:::tip
-You can use the `uorb top [<message_name>]` command to verify in real-time that your message is published and the rate (see [uORB Messaging](../middleware/uorb.md#uorb-top-command)). 
-This approach can also be used to test incoming messages that publish a uORB topic (for other messages you might use `printf` in your code and test in SITL).
+### Streaming on Request
 
-To see the message on *QGroundControl* you will need to [build it with your MAVLink library](https://dev.qgroundcontrol.com/en/getting_started/), and then verify that the message is received using [MAVLink Inspector Widget](https://docs.qgroundcontrol.com/master/en/app_menu/mavlink_inspector.html) (or some other MAVLink tool).
-:::
+Some messages are only needed once, when particular hardware is connected, or under other circumstances.
+In order to avoid clogging communications links with messages that aren't needed you may not stream all messages by default, even at low rate.
 
-## Receiving Custom MAVLink Messages
+If you needed, a GCS or other MAVLink API can request that particular messages are streamed at a particular rate using [MAV_CMD_SET_MESSAGE_INTERVAL](https://mavlink.io/en/messages/common.html#MAV_CMD_SET_MESSAGE_INTERVAL).
+A particular message can be requested just once using [MAV_CMD_REQUEST_MESSAGE](https://mavlink.io/en/messages/common.html#MAV_CMD_REQUEST_MESSAGE).
+
+## Receiving MAVLink Messages
 
 This section explains how to receive a message over MAVLink and publish it to uORB.
 
-Add a function that handles the incoming MAVLink message in
-[mavlink_receiver.h](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/mavlink_receiver.h#L77)
+It assumes that we are receiving the `BATTERY_STATUS_DEMO` message and we want to update the (existing) [BatteryStatus uORB message](../msg_docs/BatteryStatus.md) with the contained information.
+This is the kind of implementation that you would provide to support a MAVLink battery integration with PX4.
 
-```C
-#include <uORB/topics/ca_trajectory.h>
-#include <v2.0/custom_messages/mavlink_msg_ca_trajectory.h>
+Add the headers for the incoming MAVLink message and the uORB topic to publish to in [mavlink_receiver.h](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/mavlink_receiver.h#L77):
+
+```cpp
+#include <uORB/topics/battery_status.h>
+#include <v2.0/development/mavlink.h>
```

Comment:
I presume that we don't need the mavlink import even for custom messages? Presumably this is done automatically based on config value?